### PR TITLE
Fix gtest-not-found bug

### LINKED ISSUE #1030: gmock issue building realtime_tools on arch linux
I'm trying to build [ros-melodic-realtime-tools](https://aur.archlinux.org/packages/ros-melodic-realtime-tools/) from the arch linux AUR. [The PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ros-melodic-realtime-tools) shows how the make commands are called.

I get the following error:
```
$ makepkg
==> Making package: ros-melodic-realtime-tools 1.15.0-0 (Mon 09 Sep 2019 01:57:27 AM CDT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found ros-melodic-realtime-tools-1.15.0.tar.gz
==> Validating source files with sha256sums...
    ros-melodic-realtime-tools-1.15.0.tar.gz ... Passed
==> Extracting sources...
  -> Extracting ros-melodic-realtime-tools-1.15.0.tar.gz with bsdtar
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- Using CATKIN_DEVEL_PREFIX: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/melodic
-- This workspace overlays: /opt/ros/melodic
-- Using PYTHON_EXECUTABLE: /usr/bin/python3
-- Using default Python package layout
-- Using empy: /usr/lib/python3.7/site-packages/em.py
-- Using CATKIN_ENABLE_TESTING: ON
-- Skip enable_testing() when building binary package
-- Using CATKIN_TEST_RESULTS_DIR: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-3.7
-- catkin 0.7.17
-- BUILD_SHARED_LIBS is on
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)


CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.


-- Configuring incomplete, errors occurred!
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeOutput.log".
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeError.log".
```

I made a slight edit to a function to see where it was looking.
```
$ cat /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake
function(_catkin_add_executable_with_google_test type target)
...
  if(NOT ${type_upper}_FOUND AND NOT ${type_upper}_FROM_SOURCE_FOUND)
    message(WARNING "skipping ${type} '${target}' in project '${PROJECT_NAME}' because ${type} was not found in '${GTEST_INCLUDE_DIRS}'")
```

This prints out as:
```
skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found in '/usr/include'
```

I do have gmock/gtest installed, and the dirs are in `/usr/include`. For the heck of it, I tried symlinking all `*.h` files into `/usr/include` directly, but that didn't work out.

```
$ pacman -Qi gmock
Name            : gmock
Version         : 1.8.1-3

$ pacman -Qi gtest
Name            : gtest
Version         : 1.8.1-3

$ ls /usr/include/gmock/
internal         gmock-cardinalities.h      gmock-generated-function-mockers.h  gmock-generated-nice-strict.h  gmock-matchers.h      gmock-more-matchers.h
gmock-actions.h  gmock-generated-actions.h  gmock-generated-matchers.h          gmock.h                        gmock-more-actions.h  gmock-spec-builders.h
```

I think this is related to #933 and #970. I also saw #1022  and adapted ros-melodic-catkin to build from @sloretz 's fork, but this did not solve my issue. I get the same original error about gmock not being found.

Any other details that would be helpful, or debug output I could add to assist with this?

### PR DESCRIPTION
**Pull request by mikepurvis**
This change fixes #1030, which arises from the changes made in #1022.

It works for me on Ubuntu 16.04, but it should be checked by a user on Debian Buster to confirm that the original fix still works as expected.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
@sloretz can you please try this with Debian Buster?

**Comment by mikepurvis** [comment]
@jwhendy Can you also check if this fixes your build issues on Arch?

File: cmake/test/gtest.cmake [hunk 1]
```diff
@@ -226,16 +227,19 @@ endfunction()
 # :param[out] gmock_main_libs: GMock's main libraries
 # :param[out] base_dir: The base directory containing Google Test and/or GMock CMakeLists.txt
 #
-function(catkin_find_google_test_source include_path src_path
+function(catkin_find_google_test_source gtest_path googletest_path
```
Comment (mikepurvis) [inline cmake/test/gtest.cmake:230]: Obviously this is changing the signature of the function, but I'm pretty sure the only place that's calling it is in this file; it's not really an external interface, so it should be safe to change.

**Comment by jwhendy** [comment]
I can confirm that the `fix-1030` branch from @mikepurvis works when building `ros-melodic-realtime-tools` on arch linux, while `kinetic-devel` does not.

I do get a bunch of warnings like this, but they don't seem to affect the build.

```
CMake Warning (dev) at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  Policy CMP0013 is not set: Duplicate binary directories are not allowed.
  Run "cmake --help-policy CMP0013" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The binary directory

    /home/jwhendy/catkin_ws/build/moveit_tutorials/gtest

  is already used to build a source directory.  This command uses it to build
  source directory

    /usr/src/gtest

  which can generate conflicting build files.  CMake does not support this
  use case but it used to work accidentally and is being allowed for
  compatibility.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

**Edit:** this is one output from my `catkin build` process, but most/all packages seem to spit it out. Just adding this as it might be confusing that `realtime-tools` was the topic of discussion and this error was for `moveit_tutorials`.

Thanks for taking a stab at this @mikepurvis !

**Comment by dirk-thomas** [comment]
@jwhendy Please provide more context: what platform are you seeing these warnings on, what version of CMake are you using? Also this seems to be unrelated to this PR. Can you please check if the same warning happens with older versions of `catkin`?

**Comment by sloretz** [comment]
> @sloretz can you please try this with Debian Buster?

I'm able to build on Debian buster up to and including `realtime_tools` (with ros-controls/realtime_tools#46) using this branch, though there are a couple extra include paths passed to g++. The ones ending in `/include` make sense to me, but the ones without it don't. I haven't figured out where they come from yet.


```
/usr/bin/c++  -DROSCONSOLE_BACKEND_LOG4CXX -DROS_BUILD_SHARED_LIBS=1 -DROS_PACKAGE_NAME=\"realtime_tools\" -I/home/sloretz/rtt/src/realtime_tools/include -I/home/sloretz/ws/noetic/devel_isolated/roscpp/include -I/home/sloretz/ws/noetic/devel_isolated/roscpp/include/ros -I/home/sloretz/ws/noetic/src/ros_comm/clients/roscpp/include -I/home/sloretz/ws/noetic/src/ros_comm/utilities/xmlrpcpp/include -I/home/sloretz/ws/noetic/src/ros_comm/utilities/xmlrpcpp/include/xmlrpcpp -I/home/sloretz/ws/noetic/devel_isolated/rosgraph_msgs/include -I/home/sloretz/ws/noetic/devel_isolated/std_msgs/include -I/home/sloretz/ws/noetic/src/std_msgs/include -I/home/sloretz/ws/noetic/src/rosconsole/include -I/home/sloretz/ws/noetic/src/roscpp_core/roscpp_serialization/include -I/home/sloretz/ws/noetic/src/roscpp_core/rostime/include -I/home/sloretz/ws/noetic/src/roscpp_core/roscpp_traits/include -I/home/sloretz/ws/noetic/src/roscpp_core/cpp_common/include -isystem /usr/src/googletest/googlemock/include -isystem /usr/src/googletest/googlemock -isystem /usr/src/googletest/googletest/include -isystem /usr/src/googletest/googletest  -O3 -DNDEBUG   -pthread -DGTEST_HAS_PTHREAD=1 -o CMakeFiles/realtime_box_tests.dir/test/realtime_box_tests.cpp.o -c /home/sloretz/rtt/src/realtime_tools/test/realtime_box_tests.cpp
```

From above, highlighting:
```
    -isystem /usr/src/googletest/googlemock/include
    -isystem /usr/src/googletest/googlemock
    -isystem /usr/src/googletest/googletest/include
    -isystem /usr/src/googletest/googletest
```

**Comment by jwhendy** [comment]
@dirk-thomas sorry on the delay, I was out of town. It is unrelated to this PR, but not unrelated to my overall issues of getting catkin and gmock to work on Arch in #1030. I forgot that I already [commented on](https://github.com/ros/catkin/issues/1030#issuecomment-539803783) this over there. 

I'll add a followup comment there instead of muddying the waters here, but this is almost certainly an issue on how to get Arch + gmock package files correct, nothing to do with catkin. Sorry about that.

**Comment by sloretz** [comment]
> there are a couple extra include paths passed to g++. The ones ending in /include make sense to me, but the ones without it don't. I haven't figured out where they come from yet.

I think this is from the CMakeLists.txt shipped with googletest, so it can be ignored.

```cmake
# Where Google Test's .h files can be found.
set(gtest_build_include_dirs
  "${gtest_SOURCE_DIR}/include"
  "${gtest_SOURCE_DIR}")
include_directories(${gtest_build_include_dirs})
```

**Review by sloretz — APPROVED** [review]
I don't understand why this works, but it works on Debian Buster. LGTM, though I haven't tried it on all platforms.

**Comment by dirk-thomas** [comment]
> I don't understand why this works,

:zipper_mouth_face: 

> but it works on Debian Buster.

:ship: :crossed_fingers:

**Comment by dirk-thomas** [comment]
Thanks to everyone for creating / testing the patch.

File: cmake/test/gtest.cmake [hunk 2]
```diff
@@ -226,16 +227,19 @@ endfunction()
...
-  set(_gtest_source_paths "${src_path}/gtest/src")
-  list(APPEND _gtest_include_paths "${include_path}/googletest/googletest/include/gtest")
-  list(APPEND _gtest_source_paths "${src_path}/googletest/googletest/src")
+  # Path to gtest from the libgtest-dev Debian package.
+  set(_gtest_include_paths "${gtest_path}/include/gtest")
+  set(_gtest_source_paths "${gtest_path}/src/gtest/src")
+
+  # Path to gtest from the googletest Debian package.
+  list(APPEND _gtest_include_paths "${googletest_path}/googletest/include/gtest")
+  list(APPEND _gtest_source_paths "${googletest_path}/googletest/googletest/src")
```
Comment (dirk-thomas) [inline cmake/test/gtest.cmake:241]: The double "googletest" seems to be a mistake. The `include` and `src` directories should be siblings. See #1088.

**Pull request merged**